contrib/metadata-model-extensions/datahub-demo-dataset-governance-validator/LOCAL_DEVELOPER_STEP_BY_STEP_GUIDE.md

```diff
@@ -0,0 +1,549 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Custom validator that ensures dataset URNs have ownership, tags, and domain set
+ * either in the current batch of MCPs or already existing in GMS.
+ */
+public class DatasetGovernanceValidator extends AspectPayloadValidator {
+
+  private static final Logger logger = LoggerFactory.getLogger(DatasetGovernanceValidator.class);
+
+  private static final String DATASET_ENTITY_TYPE = "dataset";
+  private static final Set<String> REQUIRED_ASPECTS = Set.of(
+      Constants.OWNERSHIP_ASPECT_NAME,      // "ownership"
+      Constants.GLOBAL_TAGS_ASPECT_NAME,    // "globalTags"
+      Constants.DOMAINS_ASPECT_NAME         // "domains"
+  );
+
+  private AspectPluginConfig config;
+
+  @Override
+  protected Stream<AspectValidationException> validateProposedAspects(
+      @Nonnull Collection<? extends BatchItem> mcpItems,
+      @Nonnull RetrieverContext retrieverContext) {
+
+    logger.debug("Validating {} MCP items", mcpItems.size());
+
+    AspectRetriever aspectRetriever = retrieverContext.getAspectRetriever();
+
+    // Group all dataset URNs in the batch
+    Set<Urn> datasetUrns = mcpItems.stream()
+        .map(BatchItem::getUrn)
+        .filter(urn -> DATASET_ENTITY_TYPE.equals(urn.getEntityType()))
+        .collect(Collectors.toSet());
+
+    if (datasetUrns.isEmpty()) {
+      return Stream.empty();
+    }
+
+    // Create map for batch fetching existing aspects
+    Map<Urn, Set<String>> urnAspects = datasetUrns.stream()
+        .collect(Collectors.toMap(
+            urn -> urn,
+            urn -> REQUIRED_ASPECTS
+        ));
+
+    // Batch fetch existing aspects from GMS
+    Map<Urn, Map<String, SystemAspect>> existingAspects =
+        aspectRetriever.getLatestSystemAspects(urnAspects);
+
+    // Track which aspects are being set in this batch
+    Map<Urn, Set<String>> batchAspects = mcpItems.stream()
+        .filter(item -> DATASET_ENTITY_TYPE.equals(item.getUrn().getEntityType()))
+        .filter(item -> REQUIRED_ASPECTS.contains(item.getAspectName()))
+        .collect(Collectors.groupingBy(
+            BatchItem::getUrn,
+            Collectors.mapping(BatchItem::getAspectName, Collectors.toSet())
+        ));
+
+    // Validate each dataset URN
+    return datasetUrns.stream()
+        .map(urn -> validateDatasetGovernance(urn, existingAspects.get(urn),
+                                              batchAspects.get(urn), mcpItems))
+        .filter(Objects::nonNull);
+  }
+
+  @Override
+  protected Stream<AspectValidationException> validatePreCommitAspects(
+      @Nonnull Collection<ChangeMCP> changeMCPs,
+      @Nonnull RetrieverContext retrieverContext) {
+
+    // Use same validation logic for pre-commit
+    return validateProposedAspects(changeMCPs, retrieverContext);
```

Comment:
Typically one or the other is needed. If the request can be validated outside of a database transactions based entirely on the request, then this is not needed. Alternatively if we need to execute in a transaction then we typically can't really validate it earlier in the request processing.